cloudstack/loadbalancer.go

```diff
@@ -843,25 +844,37 @@ func sortPorts(ports []v1.ServicePort) {
 	})
 }
 
-func comparePorts(ports []v1.ServicePort, rule *loadBalancerRule) bool {
+func comparePorts(service *v1.Service, rule *loadBalancerRule) bool {
+	ports := service.Spec.Ports
+	_, useTargetPort := getLabelOrAnnotation(service.ObjectMeta, lbUseTargetPort)
 	sortPorts(ports)
 	var additionalPorts []string
 	for _, p := range ports[1:] {
-		additionalPorts = append(additionalPorts, fmt.Sprintf("%d:%d", p.Port, p.NodePort))
+		if useTargetPort {
+			additionalPorts = append(additionalPorts, fmt.Sprintf("%d:%d", p.Port, p.Port))
+		} else {
+			additionalPorts = append(additionalPorts, fmt.Sprintf("%d:%d", p.Port, p.NodePort))
+		}
 	}
 	sort.Strings(additionalPorts)
 	sort.Strings(rule.AdditionalPortMap)
 	if len(rule.AdditionalPortMap) == 0 {
 		rule.AdditionalPortMap = nil
 	}
+	if useTargetPort {
+		return reflect.DeepEqual(additionalPorts, rule.AdditionalPortMap) &&
+			rule.Privateport == strconv.Itoa(int(ports[0].Port)) &&
+			rule.Publicport == strconv.Itoa(int(ports[0].Port))
+	}
```

Comment:
I think this could be replaced with something like:
```
privatePort := ports[0].NodePort
if useTargetPort {
	privatePort = ports[0].TargetPort
}
```
And then use `privatePort` on the return below instead of `ports[0].NodePort`.